File: projects/GPL-FH-CSharp/features/WeightedWithNeighbors/Graph.cs
//layer Weighted;

// ***********************************************************************
 
public class Graph {//refines
    // Adds an edge with weights
    public void AddAnEdge( Vertex start,  Vertex end, int weight )
    {
        AddEdge( start, new  Neighbor( end, weight ) );
    }
      
    public void AddEdge( Vertex start,  Neighbor theNeighbor )
    {
        original( start,theNeighbor );
          
        // At this point the edges are Added.
        // If there is an adorn like weight it has to be Added to
        // the neighbor already present there
        if ( isDirected==false ) 
      {
            // It has to Add ONLY the weight object to the neighbor
            Vertex end = theNeighbor.neighbor;
            end.AddWeight( end,theNeighbor.weight );
        
        } // of else
    }
    
    public void display() 
    {
        original();
    }

}

File: plugins/de.ovgu.featureide.examples/featureide_examples/GPL-FH-JML/features/WeightedOnlyVertices/GPL/Graph.java
package GPL;

import java.util.LinkedList;

// ************************************************************
 
public class Graph {
    // Adds an edge with weights
	/*@requires start != null && end != null && weight != null;@*/
	/*@ensures findsEdge(start,end) != null && findsEdge(start, end).getWeight()==weight;@*/
    public void addAnEdge( Vertex start,  Vertex end, int weight )
   {
        addEdge( start,end, weight );
    }
    /*@requires start != null && end != null ;@*/
	/*@ ensures \result.getOtherVertex(start)==end && \result.getOtherVertex(end)==start; @*/
    /*@ensures !isDirected ==> (theNeighbor.neighbor.getWeight()==theNeighbor.weight);@*/
    public void addEdge( Vertex start,  Vertex end, int weight )
   {
        addEdge( start,end ); // adds the start and end as adjacent
        start.addWeight( weight ); // the direction layer takes care of that
                
        // if the graph is undirected you have to include 
        // the weight of the edge coming back
        if ( isDirected==false )
            end.addWeight( weight );
    }
    
    public void display() 
   {
        original();
    }

}
